daemon/SecretsImpl/dataprotector.cpp

```diff
@@ -0,0 +1,326 @@
+            maxOccourenceIndex = i;
+        }
+    }
+    int minimumValidOccourences = fileContents.size() / 2 + 1;
+    if (occourences[maxOccourenceIndex] < minimumValidOccourences) {
+        qWarning(lcSailfishSecretsDaemon) << "Could not decide which file is correct, there is no majority.";
+        return Irretrievable;
+    }
+
+    m_data = fileContents[maxOccourenceIndex];
+    *result = m_data;
+
+    return Success;
+}
+
+DataProtector::Status DataProtector::putData(const QByteArray &bytes)
+{
+    QDir dir(m_path);
+
+    // Check if the main directory exists, create if not
+    if (!dir.exists()) {
+        if (!dir.mkpath(m_path)) {
+            qCWarning(lcSailfishSecretsDaemon) << "Can't create protected root directory:" << m_path;
+            return ErrorCannotCreateDirectory;
+        }
+
+        qCDebug(lcSailfishSecretsDaemon) << "Protected root directory didn't exist, so the data is empty.";
+    }
+
+    QFileInfoList oldDirectories = dir.entryInfoList(QDir::Dirs | QDir::NoDotAndDotDot);
+
+    QString dataDirName = QUuid::createUuid().toString().remove('{').remove('}');
+    QString dataDirPath = m_path + QStringLiteral("/") + dataDirName;
+
+    if (!dir.mkdir(dataDirPath)) {
+        qCWarning(lcSailfishSecretsDaemon) << "Can't create data directory:" << dataDirPath;
+        return ErrorCannotCreateDirectory;
+    }
+
+    QVector<QString> dataFilePaths = {
+        dataDirPath + QStringLiteral("/file0"),
+        dataDirPath + QStringLiteral("/file1"),
+        dataDirPath + QStringLiteral("/file2"),
+    };
+
+    QVector<QSharedPointer<QSaveFile> > files;
+    std::transform(dataFilePaths.begin(), dataFilePaths.end(), std::back_inserter(files), [](const QString &filePath) {
+        return QSharedPointer<QSaveFile>(new QSaveFile(filePath));
+    });
+
+    bool allOpen = std::all_of(files.begin(), files.end(), [](QSharedPointer<QSaveFile> &file) {
+        bool isOpen = file->open(QIODevice::WriteOnly);
+        if (!isOpen) {
+            qCWarning(lcSailfishSecretsDaemon) << "Can't open file for writing:" << file->fileName();
+        }
+        return isOpen;
+    });
+
+    if (!allOpen) {
+        qCWarning(lcSailfishSecretsDaemon) << "Not all files could be open for writing.";
+        return ErrorCannotOpenFile;
+    }
+
+    for (QSharedPointer<QSaveFile> &file : files) {
+        qint64 bytesWritten = file->write(bytes);
+        if (bytesWritten != bytes.size()) {
+            qCWarning(lcSailfishSecretsDaemon) << "Can't write file:" << file->fileName();
+            return ErrorCannotWriteFile;
+        }
+        file->commit();
+    }
+
+    // Remove old data directories
+    for (QFileInfo &fileInfo : oldDirectories) {
+        QDir oldDataDir(fileInfo.absoluteFilePath());
+        oldDataDir.removeRecursively();
+    }
+
+    // Clear in-memory data, so it gets refreshed on next call
+    m_data.clear();
+    return Success;
+}
+
+
+DataProtector::Status DataProtector::processLegacy()
```

Comment:
This feels a bit wrong to me as we have not released this yet. Hence, I'd just have a disclaimer in commit message that  old legacy files must be nuked e.i. rm -rf privileged/Secrets

Reply:
Okay, I can get rid of this.

Reply:
I think the fact that it's a directory is just an implementation detail. Thus I would advise not to have the calling code depend on that detail. This data protection solution is only what we could come up with in such a short notice, but we will refine it when/if we have a better idea and time to implement it.